src/main/java/org/jenkinsci/plugins/workflow/cps/CpsThread.java

```diff
@@ -156,7 +158,7 @@ public StepExecution getStep() {
         final CpsThread old = CURRENT.get();
         CURRENT.set(this);
 
-        try {
+        try (Timeout timeout = Timeout.limit(5, TimeUnit.MINUTES)) {
```

Comment:
Safe to do because any long-running step should be async anyway, and if your next chunk of groovy needs 5 minutes to execution something is horribly, horribly wrong. 

Note: potential for breakage here if there's a valid case where that could happen without user error (or users do something wrong frequently and unknowingly).  If that sparks any sudden concerns, consider another approach.